embassy-stm32/src/timer/ll_async.rs

```diff
@@ -0,0 +1,95 @@
+//! Low level async timer driver.
+
+use core::future::Future;
+use core::marker::PhantomData;
+use core::pin::Pin;
+use core::task::{Context, Poll};
+
+use super::{Channel, GeneralInstance4Channel};
+
+/// All timer interrupts
+#[derive(Clone, Copy)]
+pub enum InterruptFlag {
+    /// Update
+    Update = 0,
+    /// Capture/compare 1
+    CaptureCompare1 = 1,
+    /// Capture/compare 1
+    CaptureCompare2 = 2,
+    /// Capture/compare 1
+    CaptureCompare3 = 3,
+    /// Capture/compare 1
+    CaptureCompare4 = 4,
+    /// COM event
+    ComEvent = 5,
+    /// Trigger
+    Trigger = 6,
+    /// Break
+    Break = 7,
+}
+
+/// Timer future
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct TimerEventFuture<T: GeneralInstance4Channel> {
+    flag: InterruptFlag,
+    phantom: PhantomData<T>,
+}
+
+impl<T: GeneralInstance4Channel> TimerEventFuture<T> {
+    /// Enable the interrupt source and returns a new instance of Future
+    pub fn new(flag: InterruptFlag) -> Self {
```

Comment:
I could be wrong. Isn't it the case that if new is pub then anyone can create a future and then drop it causing that interrupt to become disabled.